pkg/corerp/backend/deployment/deploymentprocessor.go

```diff
@@ -575,18 +574,13 @@ func (dp *deploymentProcessor) getResourceDataByID(ctx context.Context, resource
 
 func (dp *deploymentProcessor) buildResourceDependency(resourceID resources.ID, applicationID string, resource v1.DataModelInterface, outputResources []rpv1.OutputResource, computedValues map[string]any, secretValues map[string]rpv1.SecretValueReference, recipeData portableresources.RecipeData) (ResourceData, error) {
 	var appID *resources.ID
-	if applicationID != "" {
-		parsedID, err := resources.ParseResource(applicationID)
-		if err != nil {
-			return ResourceData{}, v1.NewClientErrInvalidRequest(fmt.Sprintf("application ID %q for the resource %q is not a valid id. Error: %s", applicationID, resourceID.String(), err.Error()))
-		}
-		appID = &parsedID
-	} else if rp_pr.IsValidPortableResourceType(resourceID.TypeSegments()[0].Type) {
-		// Application id is optional for portable resource types
-		appID = nil
-	} else {
-		return ResourceData{}, fmt.Errorf("missing required application id for the resource %q", resourceID.String())
+	// This code path is used only by core application resources.
+	// Application ID is a required property for these resources and cannot be empty.
+	parsedID, err := resources.ParseResource(applicationID)
```

Comment:
I'm a bit confused.. `buildResourceDependency` is used for non-core resources as well: https://github.com/radius-project/radius/blob/0ac499f38d53497abe20a24602accfef7d33fabd/pkg/corerp/backend/deployment/deploymentprocessor.go#L516-L569

The function that we discussed is only called for core resources is `getApplicationAndEnvironmentForResourceID`
https://github.com/radius-project/radius/blob/0ac499f38d53497abe20a24602accfef7d33fabd/pkg/corerp/backend/deployment/deploymentprocessor.go#L216